Lasagna (basics) feedback

Hi! I'm going through the very first learning exercise on the track as a student who has never written a single line of Golang before. I managed to get through it on my own without cheating, so that's great :) but I still have some suggestions for improvements.

Below might be a lot text, but the kind of changes that I am suggesting would be actually tiny, like adding an extra word here and there, or a single sentence.

## 1. Functions: multiple return values?
> Functions may have multiple (explicitly typed) return values. 

This sounds mindblowing to me. How can a single function have multiple return values? I have never seen this before. Or am I misinterpreting things and does this mean only that the type definition of the return value allows something like a union type? Anyway, my problem here is that there is no follow up to that statement. No examples of a function with "multiple return values", whatever that means in practice. Is it even relevant to the exercise? If yes, maybe a code example would be nice to have. If not, maybe it doesn't need to say anything? I have no idea. I was just confused by this sentence.

## 2. Lowercase names vs uppercase names

In the example code, a variable's name is lowercase while a constant's name is uppercase. Is this a requirement? Is it just a convention? Is it just like with function names, that both variable names and constant names can be either lowercase or uppercase and that changes whether they are public or not? I wish there was at least one sentence for each explaining the name casing, like there is for functions.

## 3. Variable names - camelCase or not?

I don't see any hints in the introduction or in the code itself as to whether I'm supposed to give variable names using camelCase or snake_case, or something else. I do have examples that constants and functions are named in PascalCase, and a variable name is supposed to be lowercase, so probably camelCase, but a specific mention of this would be nice for a total Golang beginner.

## 4. Types? I don't know them... 

So, I need to define the functions from zero, there are no function stubs. I don't know any types in Golang beyond `string` and `int` that are in the code examples in the introduction.

When I defined the function without arguments as a start, I got this compilation error:

```
# lasagna [lasagna.test]
./lasagna_test.go:39:31: too many arguments in call to RemainingOvenTime
	have (number)
	want ()
./lasagna_test.go:39:31: RemainingOvenTime(25) used as value
./lasagna_test.go:63:29: too many arguments in call to PreparationTime
	have (int)
	want ()
./lasagna_test.go:63:29: PreparationTime(tt.layers) used as value
./lasagna_test.go:88:25: too many arguments in call to ElapsedTime
	have (int, int)
	want ()
```

From this compilation error, I was convinced that `number` is the correct type of the argument in `RemainingOvenTime`, but that was wrong. They're all supposed to be `int`s. Could the instructions maybe say that? "...function that takes the actual minutes the lasagna has been in the oven (`int`)"? 

Why is this compilation error saying `have (number)` and not `have (int)` for `RemainingOvenTime`?

---

That's all my thoughts! I hope it's not too much. I can probably answer a lot of those questions myself by googling, but I wanted to share this because I think a lot of Golang beginners might have similar doubts.

--- a/concepts/basics/about.md
+++ b/concepts/basics/about.md
@@ -12,6 +12,7 @@ A package is a collection of source files located in the same directory.
 All source files in a directory must share the same package name.
 It is conventional for the package name to be the last directory in the import path. For example, the files in the ["math/rand" package](https://golang.org/src/math/rand/) begin with the statement `package rand`.
 When a package is imported, only entities (functions, types, variables, constants) who's name starts with a capital letter can be used / accessed.
+The recommended style of naming in Go is that entites will be named using `camelCase`, execpt for those meant to be accessible across packages which should be `CamelCase`.
 
 ```go
 package lasagna

--- a/concepts/basics/about.md
+++ b/concepts/basics/about.md
@@ -21,7 +21,6 @@ package lasagna
 Go functions accept zero or more parameters.
 Parameters must be explicitly typed, there is no type inference.
 
-Functions may have multiple (explicitly typed) return values.
 Values are returned from functions using the `return` keyword. 
 
 A function is invoked by specifying the function name and passing arguments for each of the function's parameters.

--- a/concepts/basics/about.md
+++ b/concepts/basics/about.md
@@ -11,6 +11,7 @@ Go applications are organized in packages.
 A package is a collection of source files located in the same directory.
 All source files in a directory must share the same package name.
 It is conventional for the package name to be the last directory in the import path. For example, the files in the ["math/rand" package](https://golang.org/src/math/rand/) begin with the statement `package rand`.
+When a package is imported, only entities (functions, types, variables, constants) who's name starts with a capital letter can be used / accessed.
 
 ```go
 package lasagna
@@ -60,8 +61,6 @@ Parameters must be explicitly typed, there is no type inference.
 Values are returned from functions using the `return` keyword.
 
 A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
-When a package is imported, only functions starting with a capital letter can
-be used / accessed.
 
 Note that Go supports two types of comments.
 Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.

--- a/concepts/basics/about.md
+++ b/concepts/basics/about.md
@@ -1,6 +1,6 @@
 # Basics About
 
-The Basics concept introduced [Go](https://golang.org) as a statically typed, compiled programming language. 
+The Basics concept introduced [Go](https://golang.org) as a statically typed, compiled programming language.
 The language is often referred to as Golang because of its domain name, golang.org, but the proper name is Go.
 
 The Basics concept and corresponding exercises introduced three major language features: Packages, Functions, and Variables.
@@ -14,34 +14,6 @@ It is conventional for the package name to be the last directory in the import p
 
 ```go
 package lasagna
-```
-
-## Functions
-
-Go functions accept zero or more parameters.
-Parameters must be explicitly typed, there is no type inference.
-
-Values are returned from functions using the `return` keyword. 
-
-A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
-When a package is imported, only functions starting with a capital letter can
-be used / accessed.
-
-Note that Go supports two types of comments.
-Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
-
-```go
-package greeting
-
-// Hello is a public function
-func Hello (name string) string {
-    return hi(name)
-}
-
-// hi is a private function
-func hi (name string) string {
-    return "hi " + name
-}
 ```
 
 ## Variables
@@ -78,4 +50,32 @@ Constants are defined using the `const` keyword and can be numbers, characters, 
 
 ```go
 const Age = 21 // Defines a numeric constant 'Age' with the value of 21
+```
+
+## Functions
+
+Go functions accept zero or more parameters.
+Parameters must be explicitly typed, there is no type inference.
+
+Values are returned from functions using the `return` keyword.
+
+A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
+When a package is imported, only functions starting with a capital letter can
+be used / accessed.
+
+Note that Go supports two types of comments.
+Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
+
+```go
+package greeting
+
+// Hello is a public function
+func Hello (name string) string {
+    return hi(name)
+}
+
+// hi is a private function
+func hi (name string) string {
+    return "hi " + name
+}
 ```

--- a/concepts/basics/introduction.md
+++ b/concepts/basics/introduction.md
@@ -22,7 +22,6 @@ package lasagna
 Go functions accept zero or more parameters.
 Parameters must be explicitly typed, there is no type inference.
 
-Functions may have multiple (explicitly typed) return values.
 Values are returned from functions using the `return` keyword. 
 
 A function is invoked by specifying the function name and passing arguments for each of the function's parameters.

--- a/concepts/basics/introduction.md
+++ b/concepts/basics/introduction.md
@@ -17,34 +17,6 @@ It is conventional for the package name to be the last directory in the import p
 package lasagna
 ```
 
-## Functions
-
-Go functions accept zero or more parameters.
-Parameters must be explicitly typed, there is no type inference.
-
-Values are returned from functions using the `return` keyword. 
-
-A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
-When a package is imported, only functions starting with a capital letter can
-be used / accessed.
-
-Note that Go supports two types of comments.
-Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
-
-```go
-package greeting
-
-// Hello is a public function
-func Hello (name string) string {
-    return hi(name)
-}
-
-// hi is a private function
-func hi (name string) string {
-    return "hi " + name
-}
-```
-
 ## Variables
 
 Go is statically-typed, which means all variables [must have a defined type](https://en.wikipedia.org/wiki/Type_system) at compile-time.
@@ -79,4 +51,32 @@ Constants are defined using the `const` keyword and can be numbers, characters, 
 
 ```go
 const Age = 21 // Defines a numeric constant 'Age' with the value of 21
+```
+
+## Functions
+
+Go functions accept zero or more parameters.
+Parameters must be explicitly typed, there is no type inference.
+
+Values are returned from functions using the `return` keyword.
+
+A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
+When a package is imported, only functions starting with a capital letter can
+be used / accessed.
+
+Note that Go supports two types of comments.
+Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
+
+```go
+package greeting
+
+// Hello is a public function
+func Hello (name string) string {
+    return hi(name)
+}
+
+// hi is a private function
+func hi (name string) string {
+    return "hi " + name
+}
 ```

--- a/concepts/basics/introduction.md
+++ b/concepts/basics/introduction.md
@@ -13,6 +13,7 @@ A package is a collection of source files located in the same directory.
 All source files in a directory must share the same package name.
 It is conventional for the package name to be the last directory in the import path. For example, the files in the ["math/rand" package](https://golang.org/src/math/rand/) begin with the statement `package rand`.
 When a package is imported, only entities (functions, types, variables, constants) who's name starts with a capital letter can be used / accessed.
+The recommended style of naming in Go is that entites will be named using `camelCase`, execpt for those meant to be accessible across packages which should be `CamelCase`.
 
 ```go
 package lasagna

--- a/concepts/basics/introduction.md
+++ b/concepts/basics/introduction.md
@@ -12,6 +12,7 @@ Go applications are organized in packages.
 A package is a collection of source files located in the same directory.
 All source files in a directory must share the same package name.
 It is conventional for the package name to be the last directory in the import path. For example, the files in the ["math/rand" package](https://golang.org/src/math/rand/) begin with the statement `package rand`.
+When a package is imported, only entities (functions, types, variables, constants) who's name starts with a capital letter can be used / accessed.
 
 ```go
 package lasagna
@@ -61,8 +62,6 @@ Parameters must be explicitly typed, there is no type inference.
 Values are returned from functions using the `return` keyword.
 
 A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
-When a package is imported, only functions starting with a capital letter can
-be used / accessed.
 
 Note that Go supports two types of comments.
 Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.

--- a/exercises/concept/lasagna/.docs/hints.md
+++ b/exercises/concept/lasagna/.docs/hints.md
@@ -4,9 +4,19 @@
 
 - An [integer value][integers] can be defined as one or more consecutive digits.
 
+If you see a `panic:` error when running the tests, this is as you haven't implemented one of the functions (it should say which one) or you have left the boilerplate in place, you need to remove the `panic(...)` line from the supplied code.
+
 ## 1. Define the expected oven time in minutes
 
 - You need to define a [constant][constants] and assign it the expected oven time in minutes.
+
+If you see an `undefined: OvenTime` error then double check that you have the constant defined.
+
+If you see an `invalid operation: got != tt.expected (mismatched types float64 and int)` error then you've put a decimal point into the `OvenTime` - remove the decimal so that it'll be an integer.
+
+If you see a `syntax error: non-declaration statement outside function body` error then this is as you forgot the `const` keyword.
+
+If you see an `syntax error: unexpected :=, expecting =` error then this is as you are trying to assign to the constant like a variable; contants are assigned using `=` not `:=`.
 
 ## 2. Calculate the remaining oven time in minutes
 

--- a/exercises/concept/lasagna/.docs/hints.md
+++ b/exercises/concept/lasagna/.docs/hints.md
@@ -4,7 +4,8 @@
 
 - An [integer value][integers] can be defined as one or more consecutive digits.
 
-If you see a `panic:` error when running the tests, this is as you haven't implemented one of the functions (it should say which one) or you have left the boilerplate in place, you need to remove the `panic(...)` line from the supplied code.
+If you see a `panic:` error when running the tests, this is because you have not implemented one of the functions (it should say which one) or you have left the boilerplate in place.
+You need to remove the `panic(...)` line from the supplied code and replace it with a real implementation.
 
 ## 1. Define the expected oven time in minutes
 
@@ -12,11 +13,12 @@ If you see a `panic:` error when running the tests, this is as you haven't imple
 
 If you see an `undefined: OvenTime` error then double check that you have the constant defined.
 
-If you see an `invalid operation: got != tt.expected (mismatched types float64 and int)` error then you've put a decimal point into the `OvenTime` - remove the decimal so that it'll be an integer.
+If you see an `invalid operation: got != tt.expected (mismatched types float64 and int)` error then you have likely put a decimal point into the `OvenTime` causing Go to infer the type as a floating point number.
+Remove the decimal and the type will be inferred as an `int`.
 
-If you see a `syntax error: non-declaration statement outside function body` error then this is as you forgot the `const` keyword.
+If you see a `syntax error: non-declaration statement outside function body` error then it is likely that you forgot the `const` keyword.
 
-If you see an `syntax error: unexpected :=, expecting =` error then this is as you are trying to assign to the constant like a variable; contants are assigned using `=` not `:=`.
+If you see a `syntax error: unexpected :=, expecting =` error then you are likely trying to assign the constant using `:=` like a variable; constants are assigned using `=` not `:=`.
 
 ## 2. Calculate the remaining oven time in minutes
 

--- a/exercises/concept/lasagna/.docs/hints.md
+++ b/exercises/concept/lasagna/.docs/hints.md
@@ -3,22 +3,15 @@
 ## General
 
 - An [integer value][integers] can be defined as one or more consecutive digits.
-
-If you see a `panic:` error when running the tests, this is because you have not implemented one of the functions (it should say which one) or you have left the boilerplate in place.
-You need to remove the `panic(...)` line from the supplied code and replace it with a real implementation.
+- If you see a `panic:` error when running the tests, this is because you have not implemented one of the functions (it should say which one) or you have left the boilerplate in place. You need to remove the `panic(...)` line from the supplied code and replace it with a real implementation.
 
 ## 1. Define the expected oven time in minutes
 
 - You need to define a [constant][constants] and assign it the expected oven time in minutes.
-
-If you see an `undefined: OvenTime` error then double check that you have the constant defined.
-
-If you see an `invalid operation: got != tt.expected (mismatched types float64 and int)` error then you have likely put a decimal point into the `OvenTime` causing Go to infer the type as a floating point number.
-Remove the decimal and the type will be inferred as an `int`.
-
-If you see a `syntax error: non-declaration statement outside function body` error then it is likely that you forgot the `const` keyword.
-
-If you see a `syntax error: unexpected :=, expecting =` error then you are likely trying to assign the constant using `:=` like a variable; constants are assigned using `=` not `:=`.
+- If you see an `undefined: OvenTime` error then double check that you have the constant defined.
+- If you see an `invalid operation: got != tt.expected (mismatched types float64 and int)` error then you have likely put a decimal point into the `OvenTime` causing Go to infer the type as a floating point number. Remove the decimal and the type will be inferred as an `int`.
+- If you see a `syntax error: non-declaration statement outside function body` error then it is likely that you forgot the `const` keyword.
+- If you see a `syntax error: unexpected :=, expecting =` error then you are likely trying to assign the constant using `:=` like a variable; constants are assigned using `=` not `:=`.
 
 ## 2. Calculate the remaining oven time in minutes
 

--- a/exercises/concept/lasagna/.docs/instructions.md
+++ b/exercises/concept/lasagna/.docs/instructions.md
@@ -18,24 +18,37 @@ OvenTime
 Define the `RemainingOvenTime()` function that takes the actual minutes the lasagna has been in the oven as a parameter and returns how many minutes the lasagna still has to remain in the oven, based on the expected oven time in minutes from the previous task.
 
 ```go
+func RemainingOvenTime(actual int) int {
+    // TODO
+}
+
 RemainingOvenTime(30)
 // Output: 10
 ```
 
 ## 3. Calculate the preparation time in minutes
 
-Define the `PreparationTime()` function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
+Define the `PreparationTime(layers int) int` function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
 
 ```go
+func PreparationTime(numberOfLayers int) int {
+    // TODO
+}
+
 PreparationTime(2)
 // Output: 4
 ```
 
 ## 4. Calculate the elapsed working time in minutes
 
-Define the `ElapsedTime()` function that takes two parameters: the first parameter is the number of layers you added to the lasagna, and the second parameter is the number of minutes the lasagna has been in the oven. The function should return how many minutes in total you've worked on cooking the lasagna, which is the sum of the preparation time in minutes, and the time in minutes the lasagna has spent in the oven at the moment.
+Define the `ElapsedTime(numberOfLayers, actualMinutesInOven int) int` function that takes two parameters: the first parameter is the number of layers you added to the lasagna, and the second parameter is the number of minutes the lasagna has been in the oven.
+The function should return how many minutes in total you've worked on cooking the lasagna, which is the sum of the preparation time in minutes, and the time in minutes the lasagna has spent in the oven at the moment.
 
 ```go
+func ElapsedTime(numberOfLayers, actualMinutesInOven int) int {
+    // TODO
+}
+
 ElapsedTime(3, 20)
 // Output: 26
 ```

--- a/exercises/concept/lasagna/.docs/instructions.md
+++ b/exercises/concept/lasagna/.docs/instructions.md
@@ -28,7 +28,7 @@ RemainingOvenTime(30)
 
 ## 3. Calculate the preparation time in minutes
 
-Define the `PreparationTime(layers int) int` function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
+Define the `PreparationTime` function that takes the number of layers you added to the lasagna as a parameter and returns how many minutes you spent preparing the lasagna, assuming each layer takes you 2 minutes to prepare.
 
 ```go
 func PreparationTime(numberOfLayers int) int {
@@ -41,7 +41,7 @@ PreparationTime(2)
 
 ## 4. Calculate the elapsed working time in minutes
 
-Define the `ElapsedTime(numberOfLayers, actualMinutesInOven int) int` function that takes two parameters: the first parameter is the number of layers you added to the lasagna, and the second parameter is the number of minutes the lasagna has been in the oven.
+Define the `ElapsedTime` function that takes two parameters: the first parameter is the number of layers you added to the lasagna, and the second parameter is the number of minutes the lasagna has been in the oven.
 The function should return how many minutes in total you've worked on cooking the lasagna, which is the sum of the preparation time in minutes, and the time in minutes the lasagna has spent in the oven at the moment.
 
 ```go

--- a/exercises/concept/lasagna/.docs/introduction.md
+++ b/exercises/concept/lasagna/.docs/introduction.md
@@ -18,7 +18,6 @@ package lasagna
 Go functions accept zero or more parameters.
 Parameters must be explicitly typed, there is no type inference.
 
-Functions may have multiple (explicitly typed) return values.
 Values are returned from functions using the `return` keyword.
 
 A function is invoked by specifying the function name and passing arguments for each of the function's parameters.

--- a/exercises/concept/lasagna/.docs/introduction.md
+++ b/exercises/concept/lasagna/.docs/introduction.md
@@ -8,6 +8,7 @@ This exercise introduces three major language features: Packages, Functions, and
 Go applications are organized in packages.
 A package is a collection of source files located in the same directory.
 All source files in a directory must share the same package name.
+When a package is imported, only entities (functions, types, variables, constants) who's name starts with a capital letter can be used / accessed.
 
 ```go
 package lasagna
@@ -57,8 +58,6 @@ Parameters must be explicitly typed, there is no type inference.
 Values are returned from functions using the `return` keyword.
 
 A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
-When a package is imported, only functions starting with a capital letter can
-be used / accessed.
 
 Note that Go supports two types of comments.
 Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.

--- a/exercises/concept/lasagna/.docs/introduction.md
+++ b/exercises/concept/lasagna/.docs/introduction.md
@@ -13,34 +13,6 @@ All source files in a directory must share the same package name.
 package lasagna
 ```
 
-## Functions
-
-Go functions accept zero or more parameters.
-Parameters must be explicitly typed, there is no type inference.
-
-Values are returned from functions using the `return` keyword.
-
-A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
-When a package is imported, only functions starting with a capital letter can
-be used / accessed.
-
-Note that Go supports two types of comments.
-Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
-
-```go
-package greeting
-
-// Hello is a public function
-func Hello (name string) string {
-    return hi(name)
-}
-
-// hi is a private function
-func hi (name string) string {
-    return "hi " + name
-}
-```
-
 ## Variables
 
 Go is statically-typed, which means all variables [must have a defined type](https://en.wikipedia.org/wiki/Type_system) at compile-time.
@@ -75,4 +47,32 @@ Constants are defined using the `const` keyword and can be numbers, characters, 
 
 ```go
 const Age = 21 // Defines a numeric constant 'Age' with the value of 21
+```
+
+## Functions
+
+Go functions accept zero or more parameters.
+Parameters must be explicitly typed, there is no type inference.
+
+Values are returned from functions using the `return` keyword.
+
+A function is invoked by specifying the function name and passing arguments for each of the function's parameters.
+When a package is imported, only functions starting with a capital letter can
+be used / accessed.
+
+Note that Go supports two types of comments.
+Single line comments are preceded by `//` and multiline comments are inserted between `/*` and `*/`.
+
+```go
+package greeting
+
+// Hello is a public function
+func Hello (name string) string {
+    return hi(name)
+}
+
+// hi is a private function
+func hi (name string) string {
+    return "hi " + name
+}
 ```

--- a/exercises/concept/lasagna/.docs/introduction.md
+++ b/exercises/concept/lasagna/.docs/introduction.md
@@ -9,6 +9,7 @@ Go applications are organized in packages.
 A package is a collection of source files located in the same directory.
 All source files in a directory must share the same package name.
 When a package is imported, only entities (functions, types, variables, constants) who's name starts with a capital letter can be used / accessed.
+The recommended style of naming in Go is that entites will be named using `camelCase`, execpt for those meant to be accessible across packages which should be `CamelCase`.
 
 ```go
 package lasagna

--- a/exercises/concept/lasagna/.meta/config.json
+++ b/exercises/concept/lasagna/.meta/config.json
@@ -5,7 +5,8 @@
   ],
   "contributors": [
     "ekingery",
-    "andrerfcsantos"
+    "andrerfcsantos",
+    "bobtfish"
   ],
   "forked_from": [
     "csharp/lucians-luscious-lasagna"

--- a/exercises/concept/lasagna/lasagna.go
+++ b/exercises/concept/lasagna/lasagna.go
@@ -2,8 +2,17 @@ package lasagna
 
 // TODO: define the 'OvenTime' constant
 
-// TODO: define the 'RemainingOvenTime()' function
+// RemainingOvenTime returns the remaining minutes based on the `actual` minutes already in the oven.
+func RemainingOvenTime(t int) int {
+	panic("RemainingOvenTime not implemented")
+}
 
-// TODO: define the 'PreparationTime()' function
+// PreparationTime calculates the time needed to prepare the lasagna based on the amount of layers.
+func PreparationTime(numberOfLayers int) int {
+	panic("PreparationTime not implemented")
+}
 
-// TODO: define the 'ElapsedTime()' function
+// ElapsedTime calculates the total time needed to create and bake a lasagna.
+func ElapsedTime(numberOfLayers, actualMinutesInOven int) int {
+	panic("ElapsedTime not implemented")
+}

